File: apps/earkart_omni/lib/models/patient/patient.model.dart
import 'dart:convert';
import 'package:earkart_omni/models/enums.dart';
import 'package:earkart_omni/models/language/language.entity.dart';
import 'package:earkart_omni/models/locations/locations.entity.dart';
import 'package:earkart_omni/models/user/user.entity.dart';
import 'patient.entity.dart';

PatientModel patientModelFromJson(String str) =>
    PatientModel.fromJson(json.decode(str));
String patientModelToJson(PatientModel data) => json.encode(data.toJson());

class PatientModel {
  final bool success;
  final String message;
  final dynamic data; // Can be PatientModelData or List<PatientModelData>

  PatientModel({required this.success, required this.message, this.data});

  factory PatientModel.fromJson(Map<String, dynamic> json) {
    var dataJson = json['data'];
    dynamic data;
    if (dataJson is List) {
      data = dataJson.map((e) => PatientModelData.fromJson(e)).toList();
    } else if (dataJson is Map<String, dynamic>) {
      data = PatientModelData.fromJson(dataJson);
    } else {
      data = null;
    }
    return PatientModel(
      success: json['success'],
      message: json['message'],
      data: data,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'success': success,
      'message': message,
      'data':
          data is List
              ? (data as List).map((e) => e.toJson()).toList()
              : data?.toJson(),
    };
  }
}

class PatientModelData extends PatientEntity {
  const PatientModelData({
    super.id,
    required super.contactNumber,
    super.code,
    required super.name,
    super.email,
    required super.gender,
    super.dob,
    super.age,
    super.password,
    required super.address,
    required super.cityId,
    required super.pincode,
    super.createdBy,
    super.updatedBy,
    super.createdAt,
    super.updatedAt,
    required super.languageId,
    required Status super.status,
    super.district,
    super.creator,
    super.updater,
    super.language,
    super.leadStatus,
    super.handledBy,
    super.districtId,
    super.stateId,
    super.countryId,
    super.city,
  });

  factory PatientModelData.fromJson(Map<String, dynamic> json) {
    return PatientModelData(
      id: json['id'],
      contactNumber: json['contactNumber'],
      code: json['code'],
      name: json['name'],
      email: json['email'],
      gender: genderFromApi(json['gender']),
      dob: json["dob"],
      age:
          json['age'] is int
              ? json['age']
              : (json['age'] != null
                  ? int.parse(json['age'].toString())
                  : null),
      password: json['password'],
      address: json['address'],
      cityId: json['cityId'],
      pincode: json['pincode'],
      createdBy: json['createdBy'],
      updatedBy: json['updatedBy'],
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
      languageId: json['languageId'],
      status: statusFromApi(json['status']),
      district:
          json['district'] != null
              ? DistrictEntity.fromJson(json['district'])
              : null,
      creator:
          json['creator'] != null ? UserEntity.fromJson(json['creator']) : null,
      updater:
          json['updater'] != null ? UserEntity.fromJson(json['updater']) : null,
      language:
          json['language'] != null
              ? LanguageEntity.fromJson(json['language'])
              : null,
      leadStatus: leadStatusFromApi(json['leadStatus']),
      handledBy: json['handledBy'],
      districtId: json['districtId'],
      stateId: json['stateId'],
      countryId: json['countryId'],
      city: json['city'] != null ? CityEntity.fromJson(json['city']) : null,
    );
  }

  @override
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'contactNumber': contactNumber,
      'code': code,
      'name': name,
      'email': email,
      'gender': gender.name.toUpperCase(),
      'dob': dob,
      'age': age,
      'password': password,
      'address': address,
      'cityId': cityId,
      'pincode': pincode,
      'createdBy': createdBy,
      'updatedBy': updatedBy,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
      'languageId': languageId,
      'status': status?.name.toUpperCase(),
      'creator': creator?.toJson(),
      'updater': updater?.toJson(),
      'language': language?.toJson(),
      'district': district?.toJson(),
      'leadStatus': toUpperSnakeCase(leadStatus?.name ?? ''),
      'handledBy': handledBy,
      'districtId': districtId,
      'stateId': stateId,
      'countryId': countryId,
      'city': city?.toJson(),
    };
  }
}

File: apps/dashboard/models/patient.model.ts
import { z } from "zod";
import { userModelDataSchema } from "./user.model";
import { LanguageModelDataSchema } from "./language.model";
import { CityModelDataSchema } from "./city.model";
import { HearingLossSeverity } from "./enums";

/** SNHL flags from API (boolean / string / number). */
const patientHearingLossBooleanSchema = z
  .union([z.boolean(), z.string(), z.number()])
  .optional()
  .nullable()
  .transform((v) => {
    if (v === undefined || v === null) return v;
    if (typeof v === "boolean") return v;
    if (typeof v === "number") return v !== 0;
    const s = String(v).toLowerCase().trim();
    if (["true", "1", "yes"].includes(s)) return true;
    if (["false", "0", "no"].includes(s)) return false;
    return undefined;
  });

export const patientModeldataSchema = z.object({
  id: z.string().optional(),
  contactNumber: z.string(),
  code: z.string().optional(),
  name: z.string(),
  email: z.string().optional().nullable(),
  gender: z.enum(["MALE", "FEMALE", "OTHER"]).optional().nullable(),
  dob: z.string().optional().nullable(),
  age: z.number().optional().nullable(),
  password: z.string().optional().nullable(),
  address: z.string().optional().nullable(),
  cityId: z.string().optional().nullable(),
  districtId: z.string().optional().nullable(),
  stateId: z.string().optional().nullable(),
  countryId: z.string().optional().nullable(),
  handledBy: z.string().optional().nullable(),
  pincode: z.string().optional().nullable(),
  createdBy: z.string().optional(),
  updatedBy: z.string().optional(),
  createdAt: z.string().optional(),
  updatedAt: z.string().optional(),
  languageId: z.string(),
  leadStatus: z.string().optional(),
  status: z.enum(["ACTIVE", "INACTIVE"]).optional(),
  creator: userModelDataSchema.optional().nullable(),
  updater: userModelDataSchema.optional().nullable(),
  language: LanguageModelDataSchema.optional().nullable(),
  city: CityModelDataSchema.optional().nullable(),
  /** SNHL — nested under patient on consultation (GET/PUT). */
  hearingLoss: patientHearingLossBooleanSchema,
  hearingLossSeverity: z
    .union([
      z.nativeEnum(HearingLossSeverity),
      z.enum(["MILD", "MODERATE", "SEVERE", "PROFOUND"]),
      z.string(),
    ])
    .optional()
    .nullable(),
}).passthrough();

export const CreatePatientRequestSchema = z.object({
  name: z.string().min(1, "Name is required"),
  contactNumber: z.string().min(1, "Contact number is required"),
  languageId: z.string().min(1, "Language is required"),
  gender: z.enum(["MALE", "FEMALE", "OTHER"]),
  email: z.string().email("Invalid email format").optional().or(z.literal("")),
  status: z.enum(["ACTIVE", "INACTIVE"]),
  password: z.string().optional(),
  countryId: z.string().min(1, "Country is required"),
  stateId: z.string().min(1, "State is required"),
  districtId: z.string().min(1, "District is required"),
  cityId: z.string().min(1, "City is required"),
  address: z.string().min(1, "Address is required"),
  pincode: z.string().min(1, "Pincode is required"),
});

export type CreatePatientRequest = z.infer<typeof CreatePatientRequestSchema>;

export const PatientResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
  data: patientModeldataSchema,
});

export type Patient = z.infer<typeof patientModeldataSchema>;

export const patientModelSchema = z.object({
  success: z.boolean(),
  message: z.string(),
  data: patientModeldataSchema.optional().nullable(),
});

export const patientListResponseSchema = z.object({
  success: patientModelSchema.shape.success,
  message: patientModelSchema.shape.message,
  data: z.array(patientModeldataSchema),
});

export type PatientListResponse = z.infer<typeof patientListResponseSchema>;

export type PatientModelData = z.infer<typeof patientModeldataSchema>;
export type PatientModel = z.infer<typeof patientModelSchema>;

